File: spotify-clonehero-next/lib/drum-transcription/export/sng.ts
/**
 * SNG binary container format writer.
 *
 * Produces a .sng file — an uncompressed container with XOR-masked file data,
 * used by Clone Hero and YARG as an alternative to .zip packaging.
 *
 * Format specification: https://github.com/mdsitton/SngFileFormat
 *
 * Browser-only: uses DataView/Uint8Array/ArrayBuffer. No Node.js Buffer.
 */

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/** Key-value metadata stored in the SNG header (replaces song.ini). */
export interface SngMetadata {
  [key: string]: string;
}

/** A file entry to include in the SNG container. */
export interface SngFileEntry {
  /** Relative filename (e.g. "notes.chart"). Must be <= 255 UTF-8 bytes. */
  filename: string;
  /** Raw file contents as Uint8Array. */
  data: Uint8Array;
}

// ---------------------------------------------------------------------------
// XOR masking
// ---------------------------------------------------------------------------

/**
 * Apply XOR masking to file data. The algorithm from the SNG spec:
 *
 * ```
 * for i = 0 to len(fileBytes) - 1:
 *     xorKey = xorMask[i % 16] XOR (i AND 0xFF)
 *     maskedBytes[i] = fileBytes[i] XOR xorKey
 * ```
 *
 * Masking is symmetric: applying the same operation to masked data
 * produces the original. The `i` counter resets to 0 for each file.
 */
export function maskFileData(
  data: Uint8Array,
  xorMask: Uint8Array,
): Uint8Array {
  const masked = new Uint8Array(data.length);
  for (let i = 0; i < data.length; i++) {
    const xorKey = xorMask[i % 16] ^ (i & 0xff);
    masked[i] = data[i] ^ xorKey;
  }
  return masked;
}

// ---------------------------------------------------------------------------
// Size calculation
// ---------------------------------------------------------------------------

interface SectionSizes {
  headerSize: number;
  metadataPayloadSize: number;
  metadataEntryCount: number;
  fileIndexPayloadSize: number;
  totalFileDataSize: number;
  totalSize: number;
  /** Absolute byte offset where the first file's data begins. */
  fileDataStartOffset: number;
  /** Filtered metadata entries (empty keys/values removed). */
  filteredEntries: [string, string][];
}

function calculateSizes(
  metadata: SngMetadata,
  files: SngFileEntry[],
): SectionSizes {
  const encoder = new TextEncoder();

  // Header: 6 (identifier) + 4 (version) + 16 (xorMask) = 26
  const headerSize = 26;

  // Metadata section: 8 (metadataLen) + 8 (metadataCount) + sum of pairs
  let metadataPayloadSize = 8; // metadataCount
  const entries = Object.entries(metadata).filter(([k, v]) => k && v);
  for (const [key, value] of entries) {
    metadataPayloadSize += 4 + encoder.encode(key).length; // keyLen + key
    metadataPayloadSize += 4 + encoder.encode(value).length; // valueLen + value
  }
  const metadataSectionSize = 8 + metadataPayloadSize; // sectionLen + payload

  // FileIndex section: 8 (fileMetaLen) + 8 (fileCount) + sum of file metas
  let fileIndexPayloadSize = 8; // fileCount
  for (const file of files) {
    fileIndexPayloadSize += 1 + encoder.encode(file.filename).length; // filenameLen + filename
    fileIndexPayloadSize += 8 + 8; // contentsLen + contentsIndex
  }
  const fileIndexSectionSize = 8 + fileIndexPayloadSize; // sectionLen + payload

  // FileData section: 8 (fileDataLen) + sum of file contents
  const totalFileDataSize = files.reduce((sum, f) => sum + f.data.length, 0);
  const fileDataSectionSize = 8 + totalFileDataSize;

  // Total
  const totalSize =
    headerSize + metadataSectionSize + fileIndexSectionSize + fileDataSectionSize;

  // The absolute offset where file data starts (right after the fileDataLen field)
  const fileDataStartOffset =
    headerSize + metadataSectionSize + fileIndexSectionSize + 8;

  return {
    headerSize,
    metadataPayloadSize,
    metadataEntryCount: entries.length,
    fileIndexPayloadSize,
    totalFileDataSize,
    totalSize,
    fileDataStartOffset,
    filteredEntries: entries,
  };
}

// ---------------------------------------------------------------------------
// BigInt helper
// ---------------------------------------------------------------------------

function setBigUint64LE(
  view: DataView,
  offset: number,
  value: bigint,
): void {
  view.setBigUint64(offset, value, true);
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Serialize files and metadata into an SNG binary container.
 * Returns the complete .sng file as a Uint8Array.
 *
 * Browser-only: uses DataView/Uint8Array, no Node.js dependencies.
 *
 * @param metadata - Key-value metadata (replaces song.ini in the SNG format).
 * @param files - File entries to include. Filenames must be <= 255 UTF-8 bytes.
 * @returns The complete .sng file as a Uint8Array.
 * @throws If any filename exceeds 255 UTF-8 bytes.
 */
export function buildSngFile(
  metadata: SngMetadata,
  files: SngFileEntry[],
): Uint8Array {
  const sizes = calculateSizes(metadata, files);
  const buffer = new ArrayBuffer(sizes.totalSize);
  const view = new DataView(buffer);
  const bytes = new Uint8Array(buffer);
  const encoder = new TextEncoder();
  let offset = 0;

  // --- Header (26 bytes) ---

  // "SNGPKG" identifier
  const identifier = encoder.encode('SNGPKG');
  bytes.set(identifier, offset);
  offset += 6;

  // version = 1
  view.setUint32(offset, 1, true);
  offset += 4;

  // xorMask: 16 random bytes
  const xorMask = crypto.getRandomValues(new Uint8Array(16));
  bytes.set(xorMask, offset);
  offset += 16;

  // --- Metadata Section ---

  // metadataLen (uint64) — byte length of everything after this field
  setBigUint64LE(view, offset, BigInt(sizes.metadataPayloadSize));
  offset += 8;

  // metadataCount (uint64)
  setBigUint64LE(view, offset, BigInt(sizes.metadataEntryCount));
  offset += 8;

  // metadata pairs
  for (const [key, value] of sizes.filteredEntries) {
    const keyBytes = encoder.encode(key);
    view.setInt32(offset, keyBytes.length, true);
    offset += 4;
    bytes.set(keyBytes, offset);
    offset += keyBytes.length;

    const valueBytes = encoder.encode(value);
    view.setInt32(offset, valueBytes.length, true);
    offset += 4;
    bytes.set(valueBytes, offset);
    offset += valueBytes.length;
  }

  // --- FileIndex Section ---

  // fileMetaLen (uint64)
  setBigUint64LE(view, offset, BigInt(sizes.fileIndexPayloadSize));
  offset += 8;

  // fileCount (uint64)
  setBigUint64LE(view, offset, BigInt(files.length));
  offset += 8;

  // file metas — contentsIndex is absolute from start of file
  let fileOffset = sizes.fileDataStartOffset;
  for (const file of files) {
    const filenameBytes = encoder.encode(file.filename);
    if (filenameBytes.length > 255) {
      throw new Error(
        `Filename "${file.filename}" exceeds 255 bytes (got ${filenameBytes.length})`,
      );
    }
    view.setUint8(offset, filenameBytes.length);
    offset += 1;
    bytes.set(filenameBytes, offset);
    offset += filenameBytes.length;
    setBigUint64LE(view, offset, BigInt(file.data.length));
    offset += 8;
    setBigUint64LE(view, offset, BigInt(fileOffset));
    offset += 8;
    fileOffset += file.data.length;
  }

  // --- FileData Section ---

  // fileDataLen (uint64)
  setBigUint64LE(view, offset, BigInt(sizes.totalFileDataSize));
  offset += 8;

  // masked file contents
  for (const file of files) {
    const masked = maskFileData(file.data, xorMask);
    bytes.set(masked, offset);
    offset += masked.length;
  }

  return bytes;
}

File: tools/youtube-align/youtube_align/sng.py
"""SNG container reader.

The .sng format is an uncompressed container with XOR-masked file data, used by
Clone Hero / YARG. This reader is the inverse of the repo's writer at
`spotify-clonehero-next/lib/chart-export/sng.ts`. Format reference:
https://github.com/mdsitton/SngFileFormat

Layout (all integers little-endian):

    Header
        6  bytes  identifier "SNGPKG"
        4  bytes  version (uint32)
        16 bytes  xorMask
    Metadata section
        8  bytes  metadataPayloadSize (uint64)
        8  bytes  metadataCount (uint64)
        repeated: int32 keyLen, key, int32 valLen, val
    FileIndex section
        8  bytes  fileIndexPayloadSize (uint64)
        8  bytes  fileCount (uint64)
        repeated: uint8 nameLen, name, uint64 contentsLen, uint64 contentsIndex
    FileData section
        8  bytes  totalFileDataSize (uint64)
        masked file bytes at the absolute `contentsIndex` offsets

Masking is symmetric (XOR), and the per-byte counter resets to 0 for each file:

    xorKey = xorMask[i % 16] ^ (i & 0xFF)
    out[i] = data[i] ^ xorKey
"""

from __future__ import annotations

import struct
from dataclasses import dataclass


@dataclass
class SngFile:
    name: str
    contents_len: int
    contents_index: int


@dataclass
class Sng:
    version: int
    metadata: dict[str, str]
    files: list[SngFile]
    _raw: bytes
    _xor_mask: bytes

    def read(self, name: str) -> bytes:
        """Return the unmasked bytes of a contained file by name."""
        for f in self.files:
            if f.name == name:
                return _unmask(
                    self._raw[f.contents_index : f.contents_index + f.contents_len],
                    self._xor_mask,
                )
        raise KeyError(name)


def _unmask(data: bytes, xor_mask: bytes) -> bytes:
    out = bytearray(len(data))
    for i in range(len(data)):
        out[i] = data[i] ^ (xor_mask[i % 16] ^ (i & 0xFF))
    return bytes(out)


def parse_sng(raw: bytes) -> Sng:
    if raw[:6] != b"SNGPKG":
        raise ValueError("not an SNG file (bad identifier)")
    off = 6
    (version,) = struct.unpack_from("<I", raw, off)
    off += 4
    xor_mask = raw[off : off + 16]
    off += 16

    # Metadata section
    (meta_payload,) = struct.unpack_from("<Q", raw, off)
    off += 8
    meta_end = off + meta_payload
    (meta_count,) = struct.unpack_from("<Q", raw, off)
    off += 8
    metadata: dict[str, str] = {}
    for _ in range(meta_count):
        (klen,) = struct.unpack_from("<i", raw, off)
        off += 4
        key = raw[off : off + klen].decode("utf-8")
        off += klen
        (vlen,) = struct.unpack_from("<i", raw, off)
        off += 4
        val = raw[off : off + vlen].decode("utf-8")
        off += vlen
        metadata[key] = val
    off = meta_end

    # FileIndex section
    (index_payload,) = struct.unpack_from("<Q", raw, off)
    off += 8
    index_end = off + index_payload
    (file_count,) = struct.unpack_from("<Q", raw, off)
    off += 8
    files: list[SngFile] = []
    for _ in range(file_count):
        nlen = raw[off]
        off += 1
        name = raw[off : off + nlen].decode("utf-8")
        off += nlen
        (clen,) = struct.unpack_from("<Q", raw, off)
        off += 8
        (cidx,) = struct.unpack_from("<Q", raw, off)
        off += 8
        files.append(SngFile(name=name, contents_len=clen, contents_index=cidx))
    off = index_end

    return Sng(
        version=version,
        metadata=metadata,
        files=files,
        _raw=raw,
        _xor_mask=xor_mask,
    )


def build_sng(metadata: dict[str, str], files: dict[str, bytes], xor_mask: bytes | None = None) -> bytes:
    """Build an SNG container. Used only by tests to exercise the reader."""
    if xor_mask is None:
        xor_mask = bytes(range(16))
    assert len(xor_mask) == 16

    meta_entries = [(k, v) for k, v in metadata.items() if k and v]
    meta_payload = 8
    for k, v in meta_entries:
        meta_payload += 4 + len(k.encode()) + 4 + len(v.encode())

    index_payload = 8
    for name in files:
        index_payload += 1 + len(name.encode()) + 8 + 8

    header_size = 26
    meta_section = 8 + meta_payload
    index_section = 8 + index_payload
    file_data_start = header_size + meta_section + index_section + 8

    out = bytearray()
    out += b"SNGPKG"
    out += struct.pack("<I", 1)
    out += xor_mask

    out += struct.pack("<Q", meta_payload)
    out += struct.pack("<Q", len(meta_entries))
    for k, v in meta_entries:
        kb, vb = k.encode(), v.encode()
        out += struct.pack("<i", len(kb)) + kb
        out += struct.pack("<i", len(vb)) + vb

    out += struct.pack("<Q", index_payload)
    out += struct.pack("<Q", len(files))
    cursor = file_data_start
    for name, data in files.items():
        nb = name.encode()
        out += struct.pack("<B", len(nb)) + nb
        out += struct.pack("<Q", len(data))
        out += struct.pack("<Q", cursor)
        cursor += len(data)

    total = sum(len(d) for d in files.values())
    out += struct.pack("<Q", total)
    for data in files.values():
        out += _unmask(data, xor_mask)  # unmask is symmetric == mask

    return bytes(out)
